File: standalone/event-ownership.js
(function(root,factory){const api=factory();if(typeof module==='object'&&module.exports)module.exports=api;else root.StudentAgeEventOwnership=api;})(typeof window==='object'?window:globalThis,()=>{
'use strict';const ids=v=>Array.isArray(v)?v.map(Number).filter(Number.isFinite):[];
function links(doc,folders){
 const edges={},talks=new Set(Object.keys(doc.talks));for(const t of talks){const r=doc.talks[t];edges[t]=[...ids(r.nextTalk),...ids(r.nextTalk2),...ids(r.option).flatMap(o=>[...ids(doc.options[o]?.talkId),...ids(doc.options[o]?.talkId2)])];}
 for(const f of Object.values(folders||{}))(edges[f.parentTalkId]??=[]).push(...ids(f.talkIds),...[f.routerId,f.exitId,f.endId].filter(v=>v!=null).map(Number));
 const reverse={};for(const [src,dests]of Object.entries(edges))for(const dest of dests)(reverse[dest]??=[]).push(Number(src));
 return {edges,reverse,talks};
}
// `talks` is the set of existing talk keys from links(); lookups through segmented-talk proxies are costly.
function walk(talks,edges,todo,assign,allow){const seen=new Set();while(todo.length){const t=todo.pop();if(seen.has(t)||!talks.has(String(t)))continue;seen.add(t);if(allow&&!allow(t))continue;assign(t);todo.push(...(edges[t]||[]));}}
function forward(talks,doc,edges){
 const reached={};
 for(const [key,e]of Object.entries(doc.events)){const id=Number(key);walk(talks,edges,[...ids(e.talkId),...ids(e.options).flatMap(o=>[...ids(doc.options[o]?.talkId),...ids(doc.options[o]?.talkId2)])],t=>(reached[t]??=new Set()).add(id));}
 return reached;
}
function listed(reached){return Object.fromEntries(Object.entries(reached).filter(([,v])=>v.size).map(([t,v])=>[t,[...v].sort((a,b)=>a-b)]));}
// Shown beside an event. Not membership: external dialogues stay in their own list.
function display(doc,folders){
 const {edges,reverse,talks}=links(doc,folders),reached=forward(talks,doc,edges),entered=new Set(Object.keys(reached).map(Number));
 const byEvent={};for(const [talk,evs]of Object.entries(reached))for(const id of evs)(byEvent[id]??=[]).push(Number(talk));
 for(const [event,members]of Object.entries(byEvent)){const id=Number(event),todo=members.slice(),seen=new Set(members),fresh=[];while(todo.length){const current=todo.pop();for(const prev of reverse[current]||[]){if(entered.has(prev)||seen.has(prev)||!talks.has(String(prev)))continue;seen.add(prev);(reached[prev]??=new Set()).add(id);todo.push(prev);fresh.push(prev);}}walk(talks,edges,fresh,t=>(reached[t]??=new Set()).add(id),t=>!(entered.has(t)&&!(reached[t]&&reached[t].has(id))));}
 // Bucket talks and options by their event-number prefix once; scanning all talks per event is quadratic on large mods.
 const talksByEvent=new Map(),optionsByEvent=new Map();
 for(const key of talks){const t=Number(key),e=Math.floor(t/1000);if(!talksByEvent.has(e))talksByEvent.set(e,[]);talksByEvent.get(e).push(t);}
 for(const [oid,row]of Object.entries(doc.options||{})){const e=Math.floor(Number(oid)/100);if(!optionsByEvent.has(e))optionsByEvent.set(e,[]);optionsByEvent.get(e).push(row);}
 for(const key of Object.keys(doc.events)){const id=Number(key);if(!id)continue;const seeds=(talksByEvent.get(id)||[]).slice();for(const row of optionsByEvent.get(id)||[])seeds.push(...ids(row.talkId),...ids(row.talkId2));walk(talks,edges,seeds,t=>(reached[t]??=new Set()).add(id),t=>{if(reached[t]&&!reached[t].has(id)&&entered.has(t)){reached[t].add(id);return false;}return true;});}
 return listed(reached);
}
function ownership(doc,folders,previous=doc.talkOwners||{},anchor=[]){
 const {edges,talks}=links(doc,folders),reached=forward(talks,doc,edges),anchored=new Set(anchor.map(Number));
 const extra=new Set(Object.keys(display(doc,folders)).map(Number).filter(t=>!reached[t]&&!anchored.has(t)));
 const retained={};for(const [t,v]of Object.entries(previous)){const id=Number(t);if(!doc.talks[t]||extra.has(id))continue;for(const e of ids(v))if(doc.events[e])(retained[e]??=[]).push(id);}
 const result={};for(const [t,v]of Object.entries(reached))result[t]=new Set(v);
 for(const key of Object.keys(doc.events)){const id=Number(key);walk(talks,edges,[...(retained[id]||[])],t=>{if(!reached[t])(result[t]??=new Set()).add(id);});}
 for(const id of anchored){const key=String(id);if(!doc.talks[key])continue;const set=result[key]??=new Set();for(const e of ids(previous[key]||previous[id]))if(doc.events[e])set.add(e);}
 return listed(result);
}
function sync(doc,folders,currentEvent,beforeIds){
 const previous={...(doc.talkOwners||{})},anchor=[];
 if(beforeIds&&doc.events[currentEvent])for(const t of Object.keys(doc.talks))if(!beforeIds.has(t)){previous[t]=[Number(currentEvent)];anchor.push(Number(t));}
 doc.talkOwners=ownership(doc,folders,previous,anchor);
}
function interactionTalks(doc){const found=new Set(),todo=Object.values(doc.interactions||{}).map(r=>Number(r.talkId)).filter(Boolean);while(todo.length){const id=todo.pop(),r=doc.talks[id];if(!r||found.has(id))continue;found.add(id);todo.push(...ids(r.nextTalk),...ids(r.nextTalk2),...ids(r.option).flatMap(o=>[...ids(doc.options?.[o]?.talkId),...ids(doc.options?.[o]?.talkId2)]));}return found;}
function plan(doc,folders,eventIds,localIds){const owners=ownership(doc,folders),removed=new Set(eventIds.map(Number)),local=new Set((localIds||[]).map(Number)),idle=interactionTalks(doc);
 const talks=new Set(Object.keys(doc.talks).map(Number).filter(t=>(owners[t]||[]).length&&(owners[t]||[]).every(e=>removed.has(e))));
 const options=new Set([...eventIds.flatMap(e=>ids(doc.events[e]?.options)),...[...talks].flatMap(t=>ids(doc.talks[t]?.option))]);for(const [id,row]of Object.entries(doc.talks))if(!talks.has(Number(id)))for(const o of ids(row.option))options.delete(o);for(const [id,row]of Object.entries(doc.events))if(!removed.has(Number(id)))for(const o of ids(row.options))options.delete(o);return {talks,options};
}
return {ownership,display,sync,plan,interactionTalks};});

File: standalone/event_ownership.py
"""Editor ownership survives disconnected runtime edges; never inferred by title."""
def ids(value):
    return [str(v) for v in value if type(v) is int or isinstance(v,str) and v.isdigit()] if isinstance(value,list) else []


def _number(value):
    return int(value) if isinstance(value,str) and value.isdigit() else None


def _links(talks, options, folders):
    edges={t:ids(r.get('nextTalk'))+ids(r.get('nextTalk2'))+[
        dest for oid in ids(r.get('option')) for field in ('talkId','talkId2') for dest in ids(options.get(oid,{}).get(field))] for t,r in talks.items()}
    for folder in folders.values():
        parent=str(folder.get('parentTalkId'))
        edges.setdefault(parent,[]).extend(ids(folder.get('talkIds'))+[str(folder[k]) for k in ('routerId','exitId','endId') if folder.get(k)])
    reverse={}
    for src,dests in edges.items():
        for dest in dests: reverse.setdefault(dest,[]).append(src)
    return edges, reverse


def _walk(talks, edges, todo, assign, allow=None):
    seen=set()
    while todo:
        t=todo.pop()
        if t in seen or t not in talks:continue
        seen.add(t)
        if allow is not None and not allow(t):continue
        assign(t);todo.extend(edges.get(t,[]))


def _forward(events, talks, options, edges):
    """Lines reached by walking forward from an event entry. This is real membership."""
    reached={}
    for event,row in events.items():
        todo=ids(row.get('talkId'))+[dest for oid in ids(row.get('options')) for field in ('talkId','talkId2') for dest in ids(options.get(oid,{}).get(field))]
        _walk(talks, edges, todo, lambda t,event=event: reached.setdefault(t,set()).add(event))
    return reached


def display_ownership(events, talks, options, folders, band=None, _prepared=None):
    """Lines to show while an event is open. This does not make them members of the event.

    A line that runs into the event, continues out of it, or sits in the event's number
    block (talk = event×1000+n, option = event×100+n) is listed with that event. It stays
    an external dialogue unless the event entry actually reaches it.
    ``band`` limits the number block to those event ids.
    """
    event_ids=set(events)
    band_ids=event_ids if band is None else {str(v) for v in band if str(v) in event_ids}
    # ownership() already has the links and forward walk; reuse them (the walk is copied, it is extended here).
    edges, reverse, forward_walk=_prepared if _prepared else (*_links(talks, options, folders), None)
    reached={t:set(v) for t,v in forward_walk.items()} if forward_walk is not None else _forward(events, talks, options, edges)
    forward=set(reached)
    by_event={}
    for talk,evs in reached.items():
        for event in evs: by_event.setdefault(event,set()).add(talk)
    for event,members in by_event.items():
        todo,seen=list(members),set(members)
        fresh=[]
        while todo:
            current=todo.pop()
            for prev in reverse.get(current,[]):
                if prev in forward or prev in seen or prev not in talks:continue
                seen.add(prev);reached.setdefault(prev,set()).add(event);todo.append(prev);fresh.append(prev)
        _walk(talks, edges, fresh, lambda t,event=event: reached.setdefault(t,set()).add(event),
              lambda t,event=event: not (t in forward and event not in reached.get(t,())))
    # Bucket talks and options by their event-number prefix once; scanning every talk for each
    # event is quadratic and took seconds on large mods.
    talks_by_event,options_by_event={},{}
    for t in talks:
        if (n:=_number(t)) is not None: talks_by_event.setdefault(n//1000,[]).append(t)
    for oid,row in options.items():
        if (n:=_number(oid)) is not None: options_by_event.setdefault(n//100,[]).append(row)
    for event in band_ids:
        number=_number(event)
        if not number:continue
        seeds=list(talks_by_event.get(number,()))
        for row in options_by_event.get(number,()):
            seeds.extend(ids(row.get('talkId'))+ids(row.get('talkId2')))
        def allow(t,event=event):
            owned=reached.get(t)
            if owned and event not in owned and t in forward:
                owned.add(event)
                return False
            return True
        _walk(talks, edges, seeds, lambda t,event=event: reached.setdefault(t,set()).add(event), allow)
    return {t:sorted(map(int,v)) for t,v in reached.items() if v}


def ownership(events, talks, options, folders, previous=None, band=None, anchor=None):
    """Real membership: the event entry's own forward chain, plus lines authored into the event.

    Lines that are only shown beside an event (a chain that runs into it, or a matching
    number with no link from the entry) stay out of this map, so the external-dialogue
    list keeps them. ``anchor`` ids were just created inside the event and stay members
    even when their number would otherwise only be a display hint.
    """
    event_ids=set(events)
    anchored={str(v) for v in (anchor or []) if str(v) in talks}
    edges,reverse=_links(talks, options, folders)
    forward=_forward(events, talks, options, edges)
    # A saved owner that exists only because an older build treated display lines as
    # members is dropped, otherwise the external list stays empty after one save.
    shown=set(display_ownership(events, talks, options, folders, band, _prepared=(edges, reverse, forward)))
    extra=shown-set(forward)-anchored
    retained={}
    for t,v in (previous or {}).items():
        if t not in talks or t in extra:continue
        for e in set(ids(v)) & event_ids: retained.setdefault(e,[]).append(t)
    result={t:set(v) for t,v in forward.items()}
    for event in events:
        _walk(talks, edges, list(retained.get(event,[])), lambda t,event=event: None if t in forward else result.setdefault(t,set()).add(event))
    for t in anchored:
        if t not in result: result[t]=set()
        for e in set(ids((previous or {}).get(t))) & event_ids: result[t].add(e)
    return {t:sorted(map(int,v)) for t,v in result.items() if v}


def deletion(events,talks,options,folders,previous,removed,interactions=None,reference_talks=None,band=None):
    owners=ownership(events,talks,options,folders,previous,band)
    removed=set(map(str,removed))
    # Remove only dialogues owned by the explicitly removed events. A line still
    # owned by a surviving event (shared continuations, merged branches) stays.
    doomed={t for t in talks if (owned := set(map(str,owners.get(t,[])))) and owned <= removed}
    option_ids={o for e in removed for o in ids(events.get(e,{}).get('options'))}
    option_ids.update(o for t in doomed for o in ids(talks[t].get('option')))
    kept_options={o for e,row in events.items() if e not in removed for o in ids(row.get('options'))}
    kept_options.update(o for t,row in talks.items() if t not in doomed for o in ids(row.get('option')))
    return doomed,option_ids-kept_options


def interaction_talks(interactions,talks,options=None):
    """Reachability from native idle chats, separate from event ownership."""
    pending=[str(r.get('talkId')) for r in interactions.values() if isinstance(r,dict) and r.get('talkId')]
    found=set();options=options or {}
    while pending:
        key=pending.pop()
        if key in found or key not in talks:continue
        found.add(key);row=talks[key]
        pending+=ids(row.get('nextTalk'))+ids(row.get('nextTalk2'))
        for oid in ids(row.get('option')):
            pending+=ids(options.get(oid,{}).get('talkId'))+ids(options.get(oid,{}).get('talkId2'))
    return found
